modules/standard/Regexp.chpl

```diff
@@ -450,7 +450,9 @@ class BadRegexpError : Error {
                    this flag can be set inside the regular expression with
                    ``(?U)``.
 
+   :throws BadRegexError: If RE2 is not created properly.
```

Comment:
So I interpreted your earlier comment (`re2 && re2->ok()`) as referring to the particular pattern being evaluated rather than the package as a whole, since `->` indicates a method call.  But here, you're referring to RE2 the third-party package.  The failure behavior of this function will depend on the arguments passed to it - I think you want to refer to the `pattern` argument here instead of the RE2 package, because pattern is the main argument sent to `qio_regexp_create_compile` to compile into a regexp object (which is what is checked by `qio_regexp_ok(ret._regexp)`).  Does that make sense?

Reply:
@lydia-duncan Mam
The argument `pattern` should be either string or byte, as mentioned. So will it be ok to consider that the procedure throws an error if the argument is not of this type?

Reply:
Unfortunately, no.  The type of the pattern argument is restricted by the [where clause](https://chapel-lang.org/docs/latest/language/spec/procedures.html#where-clauses) - if the pattern argument doesn't match the types listed there, that will result in a compilation error because we don't have an alternate function to call.  I can definitely see how you got confused, though!  The throws keyword is right before the where clause syntactically, so I can see thinking they are related.  But no, what is thrown from a function is more tied to `throw` in the *body* of the function - the `throws` keyword in the declaration is merely to indicate that it can throw, not what it throws (which is why this documentation update is so important)

Reply:
***
// Converts from Regexp error code to RE2 error code.
// Maybe some day they will diverge.  In any event, this
// hides the existence of Regexp from RE2 users.
static RE2::ErrorCode RegexpErrorToRE2(re2::RegexpStatusCode code) {
  switch (code) {
    case re2::kRegexpSuccess:
      return RE2::NoError;
    case re2::kRegexpInternalError:
      return RE2::ErrorInternal;
    case re2::kRegexpBadEscape:
      return RE2::ErrorBadEscape;
    case re2::kRegexpBadCharClass:
      return RE2::ErrorBadCharClass;
    case re2::kRegexpBadCharRange:
      return RE2::ErrorBadCharRange;
    case re2::kRegexpMissingBracket:
      return RE2::ErrorMissingBracket;
    case re2::kRegexpMissingParen:
      return RE2::ErrorMissingParen;
    case re2::kRegexpUnexpectedParen:
      return RE2::ErrorUnexpectedParen;
    case re2::kRegexpTrailingBackslash:
      return RE2::ErrorTrailingBackslash;
    case re2::kRegexpRepeatArgument:
      return RE2::ErrorRepeatArgument;
    case re2::kRegexpRepeatSize:
      return RE2::ErrorRepeatSize;
    case re2::kRegexpRepeatOp:
      return RE2::ErrorRepeatOp;
    case re2::kRegexpBadPerlOp:
      return RE2::ErrorBadPerlOp;
    case re2::kRegexpBadUTF8:
      return RE2::ErrorBadUTF8;
    case re2::kRegexpBadNamedCapture:
      return RE2::ErrorBadNamedCapture;
  }
  return RE2::ErrorInternal;
}
***
Are these error codes of some use? I think we can interpret some possible reasons for the error. 
But it might be difficult to write them down into the documentation.

Reply:
> I have read the modules for re2 and regexp, studied the function definition of the function we are working with.
An error is thrown if re2 && re2->ok() evaluates to false. In this file, in line 1022, the error statement is Invalid RE2. So should this be the reason for throwing an error?
In my opinion, this error is related to the improper creation of re2, as mentioned here

So going on with this documentation, I think that the function `ok()` returns `noError` if `re2` is created properly, else if it evaluates to `false`, one of the above error codes should be returned. So we should make use of these error codes in the comment.
What do you say about this?

Reply:
> Are these error codes of some use?

It looks like they are used by the RE2 implementation, but that the individual errors are converted into the message field for the Chapel-level `BadRegexpError` class.  So I don't think we need to go into them - if we move to having all those be individual Chapel Error subclasses, then the person making that change can update the documentation to cover them.

Reply:
I meant that we can use these errors to put some kind of constraints on our argument `pattern`. And then add it to the documentation.

Reply:
We could, though there are a lot of them so I would worry about the documentation feeling a bit bloated.  Especially because the module documentation contains a lot of information about how to specify the regular expressions already.  I'd lean towards not doing so for now and we can come back to it later if someone really wants it.

Reply:
@lydia-duncan 
Can you please tell me what could be a possible reason for this function to throw an error. I have gone through multiple files related to regular expression but couldn't find the proper explanation for this.

Reply:
You're most of the way there - I was worried listing the RE2 error codes would be too detailed, how would you generalize it?

Reply:
I am thinking of something like
```
/*
:throws BadRegexError: If the argument passed has syntactical errors.
Refer to RE2 Error Codes for more details.
*/
```
In this way, we can tell the user what could be the possible reason and where to look to solve that error.
Let me know what you think of this.

Reply:
I like it!  Is there a place with all the error codes that we can link to?  (Such as a link to a page from the RE2 implementation?)

Reply:
>static RE2::ErrorCode RegexpErrorToRE2(re2::RegexpStatusCode code) {
switch (code) {
case re2::kRegexpSuccess:
return RE2::NoError;
case re2::kRegexpInternalError:
return RE2::ErrorInternal;
case re2::kRegexpBadEscape:
return RE2::ErrorBadEscape;
case re2::kRegexpBadCharClass:
return RE2::ErrorBadCharClass;
case re2::kRegexpBadCharRange:
return RE2::ErrorBadCharRange;
case re2::kRegexpMissingBracket:
return RE2::ErrorMissingBracket;
case re2::kRegexpMissingParen:
return RE2::ErrorMissingParen;
case re2::kRegexpUnexpectedParen:
return RE2::ErrorUnexpectedParen;
case re2::kRegexpTrailingBackslash:
return RE2::ErrorTrailingBackslash;
case re2::kRegexpRepeatArgument:
return RE2::ErrorRepeatArgument;
case re2::kRegexpRepeatSize:
return RE2::ErrorRepeatSize;
case re2::kRegexpRepeatOp:
return RE2::ErrorRepeatOp;
case re2::kRegexpBadPerlOp:
return RE2::ErrorBadPerlOp;
case re2::kRegexpBadUTF8:
return RE2::ErrorBadUTF8;
case re2::kRegexpBadNamedCapture:
return RE2::ErrorBadNamedCapture;
}
return RE2::ErrorInternal;
}

These error codes are mentioned  [here](https://github.com/chapel-lang/chapel/blob/master/third-party/re2/re2-src/re2/re2.cc#L70) in Chapel Code. Can this be used for our purpose?

Reply:
I was hoping more for a webpage hosted by RE2

Reply:
Then we can refer to the `re2` repository on Github [here](https://github.com/google/re2/blob/master/re2/re2.h).

Reply:
Yeah, that looks fine to me